File: c/sieve/sieve.c
#include "sieve.h"

uint32_t sieve(uint32_t limit, uint32_t *primes, size_t max_primes) {
    uint32_t prime_counter = 0;

    if (limit < 2 || max_primes == 0) {
        return prime_counter;
    }

    // bool *is_prime = calloc(limit+1, sizeof(bool));
    bool is_prime[limit + 1];
    memset(is_prime, true, sizeof(is_prime));

    for (uint32_t i = 2; i <= limit; i++) {
        if (is_prime[i]) {
            primes[prime_counter++] = i;

            if (prime_counter >= max_primes) {
                break;
            }

            for (uint32_t j = 2; i * j <= limit; j++) {
                is_prime[i * j] = false;
            }
        }
    }

    return prime_counter;
}

File: ruby/sieve/sieve.rb
# frozen_string_literal: false

# https://exercism.org/tracks/ruby/exercises/sieve
# Sieve exercise
class Sieve
  def initialize(limit)
    @limit = limit
  end

  def primes
    result = []
    return result if @limit < 2

    result.push(2)
    return result if @limit == 2

    @numbers = Array.new(@limit + 1, true)
    @numbers.each_index { |number| @numbers[number] = false if number.modulo(2).zero? }
    @numbers[0] = @numbers[1] = false
    @numbers[2] = true

    (3..@limit).step(2).each_entry do |number|
      is_prime = @numbers[number]
      next unless is_prime

      jdex = 3
      while (mul = number * jdex) <= @limit
        index = mul
        @numbers[index] = false
        jdex += 1
      end
    end

    result = []
    @numbers.each_with_index do |is_prime, index|
      result.push(index) if is_prime
    end

    result
  end
end
